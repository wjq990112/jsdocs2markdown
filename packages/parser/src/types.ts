export interface Comp {
  name: string;
  props: Prop[];
  methods: Method[];
  description?: string;
  subComponent?: Omit<Comp, 'subComponent'>;
}

export interface Prop<T extends unknown = unknown> {
  name: string;
  type: string;
  required: boolean;
  description?: string;
  default?: T;
}

export interface Method {
  name: string;
  params: string;
  returns: string;
  required: boolean;
  description?: string;
}

export interface ParserResult {
  [key: 'default' | string]: Comp;
}
