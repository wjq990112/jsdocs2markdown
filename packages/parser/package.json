{
  "name": "@jsxdocgen/parser",
  "version": "1.0.0",
  "description": "",
  "author": "J4ck W4n9<1163585385@qq.com>",
  "scripts": {
    "test": "vitest",
    "test:cov": "vitest run --coverage"
  },
  "license": "MIT",
  "devDependencies": {
    "@types/react": "^18.0.28",
    "ts-morph": "^17.0.1",
    "vitest": "^0.28.4"
  }
}
