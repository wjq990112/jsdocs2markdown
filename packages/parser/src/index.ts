import { parser as compParser } from './comp';
// import { parser as funcParser } from './func';
export * from './comp';
// export * from './func';

export default function parser(source: string) {
  return compParser(source);
}
