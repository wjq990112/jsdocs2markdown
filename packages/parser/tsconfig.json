{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "composite": true,
    "baseUrl": ".",
    "types": ["vitest/globals"],
    "paths": {
      "@": ["src"],
      "@/*": ["src/*"]
    }
  },
  "include": ["src/**/*", "tests/**/*"]
}
