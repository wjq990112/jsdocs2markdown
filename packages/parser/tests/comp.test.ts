import { join } from 'node:path';
import read from './read';
import parser from '@';
import { type ParserResult, type Prop } from '@/types';

function readComp(path: string) {
  return read(join('./components', path));
}

describe('components with nothing', () => {
  it('should get component name from FunctionDeclaration', () => {
    const source = readComp('FunctionDeclaration.tsx');
    expect(parser(source)).toMatchSnapshot();
  });

  it('should get component name from VariableDeclaration', () => {
    const source = readComp('VariableDeclaration.tsx');
    expect(parser(source)).toMatchSnapshot();
  });
});

describe('components with props', () => {
  it('should get component props from FunctionDeclaration', () => {
    const source = readComp('FunctionDeclarationWithProps.tsx');
    expect(parser(source)).toMatchSnapshot();
  });
});
