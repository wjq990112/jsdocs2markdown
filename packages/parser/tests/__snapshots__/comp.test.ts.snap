// Vitest Snapshot v1

exports[`components with nothing > should get component name from FunctionDeclaration 1`] = `
{
  "Normal": {
    "methods": [],
    "name": "Normal",
    "props": [],
  },
  "default": {
    "methods": [],
    "name": "Default",
    "props": [],
  },
}
`;

exports[`components with nothing > should get component name from VariableDeclaration 1`] = `
{
  "Normal": {
    "methods": [],
    "name": "Normal",
    "props": [],
  },
  "default": {
    "methods": [],
    "name": "Default",
    "props": [],
  },
}
`;

exports[`components with props > should get component props from FunctionDeclaration 1`] = `
{
  "Extends": {
    "methods": [
      {
        "name": "onClick",
        "params": "evt: MouseEvent
index: number",
        "required": true,
        "returns": "void",
      },
      {
        "name": "onDoubleClick",
        "params": "evt: MouseEvent",
        "required": false,
        "returns": "void",
      },
    ],
    "name": "Extends",
    "props": [
      {
        "name": "required",
        "required": true,
        "type": "true",
      },
      {
        "name": "notRequired",
        "required": false,
        "type": "false",
      },
    ],
  },
  "Named": {
    "methods": [],
    "name": "Named",
    "props": [
      {
        "name": "required",
        "required": true,
        "type": "true",
      },
      {
        "name": "notRequired",
        "required": false,
        "type": "false",
      },
    ],
  },
  "Normal": {
    "methods": [],
    "name": "Normal",
    "props": [
      {
        "name": "required",
        "required": true,
        "type": "true",
      },
      {
        "name": "notRequired",
        "required": false,
        "type": "false",
      },
    ],
  },
  "default": {
    "methods": [],
    "name": "Default",
    "props": [
      {
        "name": "required",
        "required": true,
        "type": "true",
      },
      {
        "name": "notRequired",
        "required": false,
        "type": "false",
      },
    ],
  },
}
`;
